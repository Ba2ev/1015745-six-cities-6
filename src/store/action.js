export const ActionType = {
  UPDATE_CITY: `city/updateCity`,
  UPDATE_SORT: `sort/updateSort`,
  TOGGLE_SORTLIST: `sort/toggleSortList`,
  UPDATE_HOVERED_OFFER_ID: `offer/updateHoveredOfferId`,
  LOAD_OFFERS: `offers/loadOffers`,
  LOAD_FAVORITES_OFFERS: `offers/loadFavoritesOffers`,
  UPDATE_FAVORITES_OFFERS: `offers/updateFavoritesOffers`,
  LOAD_OFFER_DATA: `offer/loadOfferData`,
  LOAD_OFFER_COMMENTS: `offer/loadOfferComments`,
  LOAD_OFFER_NEARBY: `offer/loadOfferNearby`,
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  UPDATE_ACCOUNT: `user/updateAccount`,
};

export const ActionCreator = {
  updateCity: (city) => ({
    type: ActionType.UPDATE_CITY,
    payload: city
  }),
  updateSort: (sortType) => ({
    type: ActionType.UPDATE_SORT,
    payload: sortType
  }),
  toggleSortList: () => ({
    type: ActionType.TOGGLE_SORTLIST,
  }),
  updateHoveredOfferId: (id) => ({
    type: ActionType.UPDATE_HOVERED_OFFER_ID,
    payload: id
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers
  }),
  loadFavoritesOffer: (offers) => ({
    type: ActionType.LOAD_FAVORITES_OFFERS,
    payload: offers
  }),
  loadOfferNearby: (nearblyOffers) => ({
    type: ActionType.LOAD_OFFER_NEARBY,
    payload: nearblyOffers
  }),
  loadOfferData: (offerData) => ({
    type: ActionType.LOAD_OFFER_DATA,
    payload: offerData
  }),
  loadOfferComments: (offerComments) => ({
    type: ActionType.LOAD_OFFER_COMMENTS,
    payload: offerComments
  }),
  updateFavoriteOffers: (offerData) => ({
    type: ActionType.UPDATE_FAVORITES_OFFERS,
    payload: offerData
  }),
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  updateAccount: (email) => ({
    type: ActionType.UPDATE_ACCOUNT,
    payload: email,
  }),
};

