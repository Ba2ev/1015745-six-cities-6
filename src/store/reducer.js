import {combineReducers} from 'redux';
import {user} from './user/user';
import {property} from './property/property';
import {places} from './places/places';

export const NameSpace = {
  PROPERTY: `PROPERTY`,
  PLACES: `PLACES`,
  USER: `USER`
};

export default combineReducers({
  [NameSpace.PROPERTY]: property,
  [NameSpace.PLACES]: places,
  [NameSpace.USER]: user,
});
