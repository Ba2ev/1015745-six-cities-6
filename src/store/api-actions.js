import {ActionCreator} from './action';
import {adaptToClient} from '../offer';
import {AuthorizationStatus} from "../const";

export const fetchOffers = () => (dispatch, _getState, api) => (
  api.get(`/hotels`)
    .then(({data}) => (data.map(adaptToClient)))
    .then((data) => dispatch(ActionCreator.loadOffers(data)))
);

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(`/login`)
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .catch(() => {})
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(`/login`, {email, password})
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .then(() => dispatch(ActionCreator.updateAccount(email)))
);

export const logout = () => (dispatch, _getState, api) => (
  api.get(`/logout`)
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.NO_AUTH)))
    .then(() => dispatch(ActionCreator.updateAccount(``)))
);
