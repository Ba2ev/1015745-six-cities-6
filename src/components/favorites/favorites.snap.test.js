import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import * as redux from 'react-redux';
import configureStore from 'redux-mock-store';
import {createMemoryHistory} from 'history';
import Favorites from './favorites';
import {NameSpace} from '../../store/reducer';

const favorites = [
  {
    "city": {
      "name": `Paris`,
      "location": {
        "latitude": 48.85661,
        "longitude": 2.351499,
        "zoom": 13
      }
    },
    "previewImage": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    "images": [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/18.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/20.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/12.jpg`
    ],
    "title": `The house among olive `,
    "isFavorite": true,
    "isPremium": true,
    "rating": 3,
    "type": `room`,
    "bedrooms": 1,
    "maxAdults": 1,
    "price": 169,
    "goods": [
      `Breakfast`,
      `Air conditioning`,
      `Laptop friendly workspace`,
      `Washer`
    ],
    "host": {
      "id": 25,
      "name": `Angelina`,
      "isPro": true,
      "avatarUrl": `img/avatar-angelina.jpg`
    },
    "description": `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    "location": {
      "latitude": 48.83861,
      "longitude": 2.350499,
      "zoom": 16
    },
    "id": 1
  },
  {
    "city": {
      "name": `Cologne`,
      "location": {
        "latitude": 50.938361,
        "longitude": 6.959974,
        "zoom": 13
      }
    },
    "previewImage": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
    "images": [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/18.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/20.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`
    ],
    "title": `Penthouse, 4-5 rooms + 5 balconies`,
    "isFavorite": true,
    "isPremium": true,
    "rating": 4.2,
    "type": `hotel`,
    "bedrooms": 4,
    "maxAdults": 6,
    "price": 182,
    "goods": [
      `Laptop friendly workspace`,
      `Washer`,
      `Breakfast`,
      `Air conditioning`
    ],
    "host": {
      "id": 25,
      "name": `Angelina`,
      "isPro": true,
      "avatarUrl": `img/avatar-angelina.jpg`
    },
    "description": `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
    "location": {
      "latitude": 50.950361,
      "longitude": 6.961974,
      "zoom": 16
    },
    "id": 3
  },
];

const mockStore = configureStore({});

jest.spyOn(redux, `useSelector`);
it(`Favorites should render correctly`, () => {
  const history = createMemoryHistory();

  const initialState = {
    [NameSpace.DATA]: {
      favorites,
    },
    [NameSpace.USER]: {
      isAuthorized: true,
    }
  };

  const {container} = render(
      <redux.Provider store={mockStore(initialState)}>
        <Router history={history}>
          <Favorites />
        </Router>
      </redux.Provider>
  );

  expect(container).toMatchSnapshot();
});
