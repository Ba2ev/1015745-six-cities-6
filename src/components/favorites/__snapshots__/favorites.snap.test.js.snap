// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should render correctly 1`] = `
<div>
  <section
    class="favorites"
  >
    <h1
      class="favorites__title"
    >
      Saved listing
    </h1>
    <ul
      class="favorites__list"
    >
      <li
        class="favorites__locations-items"
      >
        <div
          class="favorites__locations locations locations--current"
        >
          <div
            class="locations__item"
          >
            <a
              class="locations__item-link"
              href="#"
            >
              <span>
                Cologne
              </span>
            </a>
          </div>
        </div>
        <div
          class="favorites__places"
        >
          <article
            class="favorites__card place-card"
            data-cart-id="3"
          >
            <div
              class="place-card__mark"
            >
              <span>
                Premium
              </span>
            </div>
            <div
              class="favorites__image-wrapper place-card__image-wrapper"
            >
              <a
                href="/offer/3"
              >
                <img
                  alt="Place image"
                  class="place-card__image"
                  height="110"
                  src="https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg"
                  width="150"
                />
              </a>
            </div>
            <div
              class="favorites__card-info place-card__info"
            >
              <div
                class="place-card__price-wrapper"
              >
                <div
                  class="place-card__price"
                >
                  <b
                    class="place-card__price-value"
                  >
                    €
                    182
                  </b>
                  <span
                    class="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  class="place-card__bookmark-button place-card__bookmark-button--active button"
                  type="button"
                >
                  <svg
                    class="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlink:href="#icon-bookmark"
                    />
                  </svg>
                  <span
                    class="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                class="place-card__rating rating"
              >
                <div
                  class="place-card__stars rating__stars"
                >
                  <span
                    style="width: 84%;"
                  />
                  <span
                    class="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                class="place-card__name"
              >
                <a
                  href="/offer/3"
                >
                  Penthouse, 4-5 rooms + 5 balconies
                </a>
              </h2>
              <p
                class="place-card__type"
              >
                hotel
              </p>
            </div>
          </article>
        </div>
      </li>
      <li
        class="favorites__locations-items"
      >
        <div
          class="favorites__locations locations locations--current"
        >
          <div
            class="locations__item"
          >
            <a
              class="locations__item-link"
              href="#"
            >
              <span>
                Paris
              </span>
            </a>
          </div>
        </div>
        <div
          class="favorites__places"
        >
          <article
            class="favorites__card place-card"
            data-cart-id="1"
          >
            <div
              class="place-card__mark"
            >
              <span>
                Premium
              </span>
            </div>
            <div
              class="favorites__image-wrapper place-card__image-wrapper"
            >
              <a
                href="/offer/1"
              >
                <img
                  alt="Place image"
                  class="place-card__image"
                  height="110"
                  src="https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg"
                  width="150"
                />
              </a>
            </div>
            <div
              class="favorites__card-info place-card__info"
            >
              <div
                class="place-card__price-wrapper"
              >
                <div
                  class="place-card__price"
                >
                  <b
                    class="place-card__price-value"
                  >
                    €
                    169
                  </b>
                  <span
                    class="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  class="place-card__bookmark-button place-card__bookmark-button--active button"
                  type="button"
                >
                  <svg
                    class="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlink:href="#icon-bookmark"
                    />
                  </svg>
                  <span
                    class="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                class="place-card__rating rating"
              >
                <div
                  class="place-card__stars rating__stars"
                >
                  <span
                    style="width: 60%;"
                  />
                  <span
                    class="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                class="place-card__name"
              >
                <a
                  href="/offer/1"
                >
                  The house among olive 
                </a>
              </h2>
              <p
                class="place-card__type"
              >
                room
              </p>
            </div>
          </article>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
